Fix #15661: no link to groups

tessi opened the pull request:
In the work package table we want to link to users, but not to groups (groups are a special kind of users).

see: https://www.openproject.org/work_packages/15661

tessi committed:
do not display links to groups in WP table

tessi committed:
test assigned_to field rendering in WP table

TeatroIO commented:
I've prepared a stage. <a href='http://fix-15661-no-link-to-groups.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Click to open.</a>

hschink commented on app/assets/javascripts/angular/work_packages/directives/work-package-column-directive.js:
```diff
@@ -125,7 +125,14 @@ angular.module('openproject.workPackages.directives')
-            return PathHelper.staticUserPath(id);
+            if (scope.workPackage[scope.column.name].type == 'Group') {
+              // if it's a group, we have nothing to link to
+              // in the (distant) future we will return a link to a group
```
The last comment line is superfluous.

tessi commented:
Thanks @myabc and @hschink for your comments.

@hschink let's talk tomorrow about this so I can be sure to understand the things you want to tell me ;)

tessi committed:
remove superfluous comment

tessi commented:
@hschink I think the PathHelper only gives a path. The decision to generate a link (no matter what path) is not done in the PathHelper. Thus, I see no point in implementing code to distinguish users and groups in the PathHelper.

The different places that generate a user/group link need to be investigated, though. (so that we can be sure not to generate group links at any other place).

tessi commented:
This are the places that request the userPath, I'll mark them when I find that they handle groups as they should:
- [x] work_packages/tabs/attachment-user-cell-directive.js:<br>
    `scope.userPath = PathHelper.staticUserPath(author.props.id);`
- [x] work_packages/tabs/user-activity-directive.js:<br>
  `scope.userPath = PathHelper.staticUserPath;`
- [x] work_packages/tabs/related-work-package-table-row-directive.js:<br>
  `scope.userPath = PathHelper.staticUserPath;`
- [x] work_packages/tabs/user_field-directive.js:<br>
  `scope.userPath = PathHelper.staticUserPath;`
- [x] work_packages/controllers/details-tab-overview-controller.js:<br>
  `$scope.userPath = PathHelper.staticUserPath;`
- [x] work_packages/controllers/work-package-details-controller.js:<br>
  `$scope.authorPath = PathHelper.staticUserPath($scope.author.props.id);`
- [x] work_packages/directives/work-package-column-directive.js:<br>
  `return PathHelper.staticUserPath(id);`

tessi committed:
don't die when the assigned_to field is not set on a WP

tessi committed:
add subtype to principals in apiv2

so we can differentiate between the different types of principals
(users vs. groups)

tessi committed:
do not render a link to a group-assigned in a related work package

tessi committed:
in WP details view do not link to groups

tessi commented:
I fixed all the templates which use the `userPath` (except those handling the author, because groups cannot be the author of something).

tessi commented:
Rebased my commits on the new `dev`

myabc commented:
As discussed with @tessi, the approach of adding `subtype` is not as clean as I would like, but good enough for the time-being. Eventually we will probably want to use some form of inheritance (inheriting from `Principal`) in the front end.
